### ISSUE TITLE
Feature request: Allocating large values on the heap (new_uninit)

### ISSUE THREAD

**Issue by hoshinolina** [Rust-for-Linux/linux#879]
The kernel has very small stacks, and `Box::try_new(T {})` does not reliably avoid putting `T {}` on the stack. It seems the only reasonable way to guarantee that `T` does not make a trip through the stack is to use unsafe code, with `Box::try_new_zeroed()` or similar. This requires the `new_uninit` feature.

I'd love to hear any better ideas of how to handle this... `try_new_zeroed` only really works safely for types where the zero bit pattern is valid, which doesn't cover all cases that `Box::try_new(T {});` does.

**Comment by nbdd0121** [comment]
Does `Box::try_new(MaybeUninit::uninit())` work for your use case? I am not aware of any issues that would prevent LLVM from optimising out the copy involved.

**Comment by hoshinolina** [comment]
The `try_new` part works, but I need to call `assume_uninit()` on the resulting Box, and that is also gated by `new_uninit`:

```rust
    #[unstable(feature = "new_uninit", issue = "63291")]
    #[rustc_const_unstable(feature = "const_box", issue = "92521")]
    #[inline]
    pub const unsafe fn assume_init(self) -> Box<T, A> {
        let (raw, alloc) = Box::into_raw_with_allocator(self);
        unsafe { Box::from_raw_in(raw as *mut T, alloc) }
    }
```

I could just duplicate the implementation, but perhaps it makes sense to just remove the `new_uninit` gate here?

**Comment by nbdd0121** [comment]
Given that:
* the main point of an unstability here is to give Rust std flexibility to change API;
* we (the kernel) control all use-sites, so we can change all use-sites if the API upstream changes

I think it's reasonable to allow this feature be used inside kernel code.

Maybe we can have a general policy to allow unstable *library* features to be used in kernel, given that it is something that can be easily implementable outside std without relying on unstable *language* features. What's your thought, @ojeda?

**Comment by bjorn3** [comment]
I agree. Although I would limit it to unstable library features that can actually be implemented outside of the standard library and don't for example depend on unstable language features themself or on implementation details of the compiler or standard library. For example `ptr::from_raw_parts` is an unstable library feature, but it depends on the exact layout of fat pointers, which is unstable. as such `ptr::from_raw_parts` shouldn't be allowed IMO.

**Comment by alyssarosenzweig** [comment]
Spectator question ... does allowing unstable library features require nightly rusts? I remember unstable language features being a pain point for using the stable rustc in Debian to build code, as opposed to something from rustup. Ideally linux would build with a sufficiently new distro rustc. Maybe library features are exempt from this issue, though? Unlike Lina I'm still a C person :(

**Comment by bjorn3** [comment]
Unstable library feature technically require nightly rustc to prevent people from depending on it in a way that we can't change them anymore, but rust-for-linux uses an escape hatch that is also used for compiling rustc itself to allow unstable features on stable and beta.

**Comment by jyn514** [comment]
> I agree. Although I would limit it to unstable library features that can actually be implemented outside of the standard library and don't for example depend on unstable language features themself or on implementation details of the compiler or standard library. For example `ptr::from_raw_parts` is an unstable library feature, but it depends on the exact layout of fat pointers, which is unstable. as such `ptr::from_raw_parts` shouldn't be allowed IMO.

At this point you may as well copy the implementation of the standard library code into your own codebase, which ensures it doesn't break in new updates, ensures you can only use the features you know about (i.e. it's not possible to use language features this way), and avoids using RUSTC_BOOTSTRAP.

**Comment by Lokathor** [comment]
You probably want to just make a function for this that wraps the unsafe part of things.

here's one way to do it: https://docs.rs/bytemuck/latest/bytemuck/allocation/fn.try_zeroed_box.html

If you don't want to take on a dep then, as jyn suggests, you could vendor the relevant code directly into your project.

**Comment by ojeda** [comment]
For this particular case, I would just enable `new_uninit`, especially since we only need the `Box` bits.

Then we can have safe methods on top too.

**Comment by ojeda** [comment]
`new_uninit` (and the pin-init API) was merged into `rust-next` -- see https://lore.kernel.org/rust-for-linux/20230408122429.1103522-1-y86-dev@protonmail.com/
